File: ruby/bob/bob.rb
class Bob

  def hey(phrase = nil)
    phrase ||= " "
    if is_yelling?(phrase)
      'Woah, chill out!'
    elsif is_question?(phrase)
      'Sure.'  
    elsif is_statement?(phrase)
      'Whatever.'
    elsif is_quiet?(phrase)
      'Fine. Be that way.'
    end
  end

  def is_question?(phrase)
    phrase.match(/.+\?$/) ? true : false
  end

  def is_yelling?(phrase)
    phrase.match(/[A-Z\W]{3,}/) &&  (phrase == phrase.upcase)

  end

  def is_statement?(phrase)
    phrase.match(/\w+/) ? true : false
  end

  def is_quiet?(phrase)
    phrase.match(/\s*/) ? true : false
  end


end

File: javascript/bob/bob.js
var Bob = function() {};

Bob.prototype.hey = function(words) {
  if (_shouting(words)) {
    return 'Whoa, chill out!';
  } if (_questioning(words)) {
    return 'Sure.';
  } if (_silent(words)) {
    return 'Fine. Be that way!';
  };
  return 'Whatever.';

  function _shouting(words) {
    return (words !== words.toLowerCase() && words === words.toUpperCase());
  };
  
  function _questioning(words) {
    return words.charAt(words.length - 1) == '?';
  };
  
  function _silent(words) {
    return /^\s*$/.test(words);
  };
};

module.exports = Bob;
